applications/callflow/src/module/cf_lookupcidname.erl

```diff
@@ -0,0 +1,120 @@
+%%% @copyright (C) 2015, 2600Hz INC
+%%% @doc
+%%% This module looks up the Caller ID Name by matching
+%%% numbers/patters with the provided lists.
+%%%
+%%% "data":{
+%%%   "lists": ["01fc63f92d9b89a25dd4ff1039e64497"]
+%%% },
+%%% @end
+%%% @contributors
+%%%     Sponsored by Conversant Ltd,
+%%%         implemented by SIPLABS, LLC (Ilya Ashchepkov)
+%%%-------------------------------------------------------------------
+-module(cf_lookupcidname).
+
+-export([handle/2]).
+
+-include("../callflow.hrl").
+
+-spec handle(wh_json:object(), whapps_call:call()) -> 'ok'.
+handle(Data, Call) ->
+    CallerIdNumber = whapps_call:caller_id_number(Call),
+    ListEntries = wh_json:get_value(<<"lists">>, Data),
+    AccountDb = whapps_call:account_db(Call),
+    lager:debug("matching ~p in ~p", [CallerIdNumber, AccountDb]),
+    CallerName = case match_prefix_in_lists(AccountDb, CallerIdNumber, ListEntries) of
+                     'continue' ->
+                         lager:debug("matching regexps"),
+                         match_regexp_in_lists(AccountDb, CallerIdNumber, ListEntries);
+                     {'stop', CallerName_} -> CallerName_
+                 end,
+    case CallerName of
+        'undefined' -> cf_exe:continue(Call);
+        CallerName ->
+            lager:info("setting caller name to ~p", [CallerName]),
+            cf_exe:continue(whapps_call:set_caller_id_name(CallerName, Call))
+    end.
+
+-spec match_prefix_in_lists(ne_binary(), ne_binary(), ne_binary() | [ne_binary()]) ->
+    {'stop', api_binary()} | 'continue'.
+match_prefix_in_lists(AccountDb, Number, Lists) ->
+    Prefixes = build_keys(Number),
+    Keys = [[ListId, Prefix] || ListId <- Lists, Prefix <- Prefixes],
+    case couch_mgr:get_results(AccountDb
+                               ,<<"lists/match_prefix_in_list">>
+                               ,[{'keys', Keys}, {'include_docs', 'true'}])
+    of
+        {'ok', [_ | _] = Matched} ->
+            lager:debug("matched ~p prefixes, getting longest", [length(Matched)]),
+            ListEntry = hd(lists:sort(fun compare_prefixes/2, Matched)),
+            Name = wh_json:get_value([<<"doc">>, <<"displayname">>], ListEntry),
+            lager:debug("matched prefix ~p", [get_prefix(ListEntry)]),
+            {'stop', Name};
+        {'ok', []} ->
+            lager:debug("no matching prefix"),
+            'continue';
+        {'error', Error} ->
+            lager:warning("error while matching prefixes: ~p", [Error]),
+            'continue'
+    end.
+
+-spec compare_prefixes(wh_json:object(), wh_json:object()) -> boolean().
+compare_prefixes(JObj1, JObj2) ->
+    byte_size(get_prefix(JObj1)) >= byte_size(get_prefix(JObj2)).
+
+-spec get_prefix(wh_json:object()) -> binary().
+get_prefix(JObj) ->
+    Doc = wh_json:get_value(<<"doc">>, JObj),
+    case wh_json:get_value(<<"number">>, Doc) of
+        'undefined' -> wh_json:get_value(<<"prefix">>, Doc, <<>>);
+        Prefix      -> Prefix
+    end.
```

Comment:
Use `get_ne_binary_value` when appropriate.
Also the `case` could be replaced using `wh_json:get_value(<<"prefix">>, Doc, <<>>)` as the default (3rd argument) in the call.